flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/operators/windowing/functions/InternalAggregateProcessAsyncWindowFunction.java

```diff
@@ -0,0 +1,114 @@
+
+/**
+ * Internal window function for wrapping a {@link ProcessWindowFunction} that takes an {@code
+ * Iterable} and an {@link AggregateFunction}. This is for async window operator.
+ *
+ * @param <K> The key type
+ * @param <W> The window type
+ * @param <T> The type of the input to the AggregateFunction
+ * @param <ACC> The type of the AggregateFunction's accumulator
+ * @param <V> The type of the AggregateFunction's result, and the input to the WindowFunction
+ * @param <R> The result type of the WindowFunction
+ */
+public final class InternalAggregateProcessAsyncWindowFunction<T, ACC, V, R, K, W extends Window>
+        extends WrappingFunction<ProcessWindowFunction<V, R, K, W>>
+        implements InternalAsyncWindowFunction<StateIterator<T>, R, K, W> {
+
+    private static final long serialVersionUID = 1L;
+
+    private final AggregateFunction<T, ACC, V> aggFunction;
+
+    public InternalAggregateProcessAsyncWindowFunction(
+            AggregateFunction<T, ACC, V> aggFunction,
+            ProcessWindowFunction<V, R, K, W> windowFunction) {
+        super(windowFunction);
+        this.aggFunction = aggFunction;
+    }
+
+    @Override
+    public StateFuture<Void> process(
+            K key,
+            final W window,
+            final InternalWindowContext context,
+            StateIterator<T> input,
+            Collector<R> out)
+            throws Exception {
+        InternalProcessWindowContext<V, R, K, W> ctx =
+                new InternalProcessWindowContext<>(wrappedFunction);
+        ctx.window = window;
+        ctx.internalContext = context;
+        AtomicReference<ACC> finalAcc = new AtomicReference<>(aggFunction.createAccumulator());
+
+        return input.onNext(
+                        val -> {
+                            finalAcc.set(aggFunction.add(val, finalAcc.get()));
+                        })
+                .thenAccept(
+                        ignore -> {
+                            ProcessWindowFunction<V, R, K, W> wrappedFunction =
+                                    this.wrappedFunction;
+                            wrappedFunction.process(
+                                    key,
+                                    ctx,
+                                    Collections.singletonList(
+                                            aggFunction.getResult(finalAcc.get())),
+                                    out);
+                        });
+    }
+
+    @Override
+    public StateFuture<Void> clear(final W window, final InternalWindowContext context)
+            throws Exception {
+        InternalProcessWindowContext<V, R, K, W> ctx =
+                new InternalProcessWindowContext<>(wrappedFunction);
+        ctx.window = window;
+        ctx.internalContext = context;
+
+        ProcessWindowFunction<V, R, K, W> wrappedFunction = this.wrappedFunction;
+        wrappedFunction.clear(ctx);
+        return StateFutureUtils.completedFuture(null);
```

Comment:
nit. Replace all `StateFutureUtils.completedFuture(null);` to `StateFutureUtils.completedVoidFuture();`

Reply:
Resolved